### PR TITLE
fix(typeahead): Typeahead uses ngSanitize when present

### PR DESCRIPTION
**Pull request by fernando-sendMail**
Hey there!

This pull request addresses the issue #2884 which mentions the use of the attribute of `bind-html-unsafe` attribute in the default `template/typeahead/typeahead-match.html`.

So basically, I changed a little the testing process, considering what @chrisirhc said about the fix, here's the implemented solution:

1. Added the `angular-sanitize.js` dependency to the `misc/test-lib/` folder (the `angular-sanitize.js` is the same version as the `angular.js` included).
2. Changed the `karma.conf.js` file to add this dependency.
3. Added a line in the demo file to ship it with the `dist` folder.
4. Changed the `typeaheadHilghlight` filter:
  * Checks if `$sanitize` is available to use, if not then a warning is printed to the console using the `$log` service, this will happen when the filer is requested by angular. The `ngSanitize` can be imported as part of the module that's using the `typeahead` directive. The reason of why the `ngSanitize` module is not imported directly on the `ui.bootstrap.typeahead` module is because it will add another dependency to the project, needless to say it will not work if not present.
  * When `$sanitize` is present it will use it to sanitize the string itself if it contains any sequence of `<` followed by `any combination of chars`, and an ending `>`, if not, then the string is wrapped in an object that `ng-bind-html` need to be displayed with the `$sce.trustAsHtml` method.
  * And finally the `<strong>` tags are added to the matched characters.

5. On the file `src/typeahead/test/typeahead-highlight.spec.js` the returned value of the filter need to be unwrapped.
6. Created another spec file (`src/typeahead/test/typeahead-highlight-ngsanitize.spec.js`) that will include the `ngSanitize` module and test that tags and attributes are being escaped.

And that's it, I think that's all.

I'm looking forward to your feedback and comments. Thanks!

### CODE REVIEW

File: src/typeahead/typeahead.js [hunk 1]
```diff
@@ -399,13 +399,30 @@ angular.module('ui.bootstrap.typeahead', ['ui.bootstrap.position', 'ui.bootstrap
...
 
     function escapeRegexp(queryToEscape) {
-      return queryToEscape.replace(/([.?*+^$[\]\\(){}|-])/g, '\\$1');
+      return queryToEscape.replace(/([.?*+^$[\]\\(){}|-])/g, '\\$1'); // Regex: capture the whole query string and replace it with the string that will be used to match the results, for example if the capture is "a" the result will be \a
     }
 
     return function(matchItem, query) {
-      return query ? ('' + matchItem).replace(new RegExp(escapeRegexp(query), 'gi'), '<strong>$&</strong>') : matchItem;
+      if($sanitize) { // Is the $sanitize service available?
+        if(/<.*>/g.test(matchItem)){
```
Comment (chrisirhc) [inline src/typeahead/typeahead.js]: I think this check isn't required. It's safe to $sanitize something even if it doesn't contain any HTML tags.

File: src/typeahead/test/typeahead-highlight.spec.js [hunk 2]
```diff
@@ -1,39 +1,50 @@
...
+    logSpy = spyOn($log, 'warn');
+  }));
+
   beforeEach(inject(function (typeaheadHighlightFilter) {
     highlightFilter = typeaheadHighlightFilter;
   }));
 
   it('should higlight a match', function () {
-    expect(highlightFilter('before match after', 'match')).toEqual('before <strong>match</strong> after');
+    expect(highlightFilter('before match after', 'match').$$unwrapTrustedValue()).toEqual('before <strong>match</strong> after');
```
Comment (chrisirhc) [inline src/typeahead/test/typeahead-highlight.spec.js]: Use `$sce.getTrustedHtml(highlightFilter('before match after', 'match'))` instead

File: src/typeahead/typeahead.js [hunk 3]
```diff
@@ -399,13 +399,30 @@ angular.module('ui.bootstrap.typeahead', ['ui.bootstrap.position', 'ui.bootstrap
...
+    try {
+      $sanitize = $injector.get('$sanitize'); //tries to inject the sanitize service
+    }
+    catch(e) {
+      $log.warn('Unsafe use of typeahead please use ngSanitize'); // Warn the user about the danger
+    }
 
     function escapeRegexp(queryToEscape) {
-      return queryToEscape.replace(/([.?*+^$[\]\\(){}|-])/g, '\\$1');
+      return queryToEscape.replace(/([.?*+^$[\]\\(){}|-])/g, '\\$1'); // Regex: capture the whole query string and replace it with the string that will be used to match the results, for example if the capture is "a" the result will be \a
```
Comment (chrisirhc) [inline src/typeahead/typeahead.js]: I think this comment can be moved to right before the declaration of function:
```
    // Regex…
    function escapeRegexp
```

File: src/typeahead/typeahead.js [hunk 4]
```diff
@@ -399,13 +399,30 @@ angular.module('ui.bootstrap.typeahead', ['ui.bootstrap.position', 'ui.bootstrap
...
     }
 
     return function(matchItem, query) {
-      return query ? ('' + matchItem).replace(new RegExp(escapeRegexp(query), 'gi'), '<strong>$&</strong>') : matchItem;
+      if($sanitize) { // Is the $sanitize service available?
+        if(/<.*>/g.test(matchItem)){
+          matchItem = $sanitize(matchItem); // If it's present the string gets sanitized and packed in a $sce object.
+        }
+      }
+      matchItem = query? ('' + matchItem).replace(new RegExp(escapeRegexp(query), 'gi'), '<strong>$&</strong>') : matchItem; // Replaces the capture string with a the same string inside of a "strong" tag
```
Comment (chrisirhc) [inline src/typeahead/typeahead.js]: The output from the `replace` isn't sanitized. Perhaps do this before calling `$sanitize` on the item.

**Comment by chrisirhc** [comment]
Did a review. Good work here. Also, PR needs rebasing.

**Comment by fernando-sendMail** [comment]
Hi @chrisirhc 
Thanks for your feedback, I polished the code according to your comments. I already pushed the commit.

**Comment by karianna** [comment]
Can't be merged at present

**Comment by fernando-sendMail** [comment]
@karianna I'm sorry, I'm kind of new to this, can I or should I do something about that?

**Comment by karianna** [comment]
@fernando-sendMail you'll need to rebase, e.g. do a `git pull angular-ui/bootstrap master` (I'm assuming you've aliased the angular-ui bootstrap remote), resolve any conflicts and `git push` into your remote branch again.

**Comment by fernando-sendMail** [comment]
@karianna Thanks for the info!

File: karma.conf.js [hunk 5]
```diff
@@ -18,6 +18,7 @@ module.exports = function(config) {
       'misc/test-lib/jquery-1.8.2.min.js',
       'node_modules/angular/angular.js',
       'node_modules/angular-mocks/angular-mocks.js',
+      'misc/test-lib/angular-sanitize.js',
```
Comment (chrisirhc) [inline karma.conf.js]: Should change this to use `node_modules/angular-sanitize/angular-sanitize.js` and modify `package.json` to include angular-sanitize as a module since we changed this in https://github.com/angular-ui/bootstrap/commit/5536e9660e6653ad5e780ce99692789d7bed6fcc so that we don't need to include the angular files in this repository.

File: misc/demo/index.html [hunk 6]
```diff
@@ -14,6 +14,7 @@
     <script src="//ajax.googleapis.com/ajax/libs/angularjs/<%= ngversion %>/angular.min.js"></script>
     <script src="//ajax.googleapis.com/ajax/libs/angularjs/<%= ngversion %>/angular-animate.min.js"></script>
     <script src="//ajax.googleapis.com/ajax/libs/angularjs/<%= ngversion %>/angular-touch.min.js"></script>
+    <script src="//ajax.googleapis.com/ajax/libs/angularjs/<%= ngversion %>/angular-sanitize.js"></script>
```
Comment (chrisirhc) [inline misc/demo/index.html]: I think you missed adding ngSanitize into the `demo/app.js`. Is this change needed for the demo to work? If not, we can just leave out this change for now and add it later when needed.

File: src/typeahead/typeahead.js [hunk 7]
```diff
@@ -399,13 +399,29 @@ angular.module('ui.bootstrap.typeahead', ['ui.bootstrap.position', 'ui.bootstrap
...
+      // Regex: capture the whole query string and replace it with the string that will be used to match 
+      // the results, for example if the capture is "a" the result will be \a
       return queryToEscape.replace(/([.?*+^$[\]\\(){}|-])/g, '\\$1');
     }
 
     return function(matchItem, query) {
-      return query ? ('' + matchItem).replace(new RegExp(escapeRegexp(query), 'gi'), '<strong>$&</strong>') : matchItem;
+      matchItem = query? ('' + matchItem).replace(new RegExp(escapeRegexp(query), 'gi'), '<strong>$&</strong>') : matchItem; // Replaces the capture string with a the same string inside of a "strong" tag
+      if($sanitize) { // Is the $sanitize service available?
+        matchItem = $sanitize(matchItem); // If it's present the string gets sanitized and packed in a $sce object.
```
Comment (chrisirhc) [inline src/typeahead/typeahead.js]: There's no need to call `$sanitize` here as ng-bind-html will automatically call it if it's not a trusted value.

File: src/typeahead/typeahead.js [hunk 8]
```diff
@@ -399,13 +399,29 @@ angular.module('ui.bootstrap.typeahead', ['ui.bootstrap.position', 'ui.bootstrap
...
 
     function escapeRegexp(queryToEscape) {
+      // Regex: capture the whole query string and replace it with the string that will be used to match 
+      // the results, for example if the capture is "a" the result will be \a
       return queryToEscape.replace(/([.?*+^$[\]\\(){}|-])/g, '\\$1');
     }
 
     return function(matchItem, query) {
-      return query ? ('' + matchItem).replace(new RegExp(escapeRegexp(query), 'gi'), '<strong>$&</strong>') : matchItem;
+      matchItem = query? ('' + matchItem).replace(new RegExp(escapeRegexp(query), 'gi'), '<strong>$&</strong>') : matchItem; // Replaces the capture string with a the same string inside of a "strong" tag
```
Comment (chrisirhc) [inline src/typeahead/typeahead.js]: Sorry, on further thought, I realized your initial logic to check for html tags should be used here like:
```js
if (containsHtml(matchItem) && !$sanitize) {
    $log.warn('Unsafe use of typeahead please use ngSanitize');
}
```

This way, the user will only be warned when using typeahead with labels that contain HTML.

File: src/typeahead/typeahead.js [hunk 9]
```diff
@@ -399,13 +399,29 @@ angular.module('ui.bootstrap.typeahead', ['ui.bootstrap.position', 'ui.bootstrap
...
 
-  .filter('typeaheadHighlight', function() {
+  .filter('typeaheadHighlight', ['$sce', '$injector', '$log', function($sce, $injector, $log) {
+
+    var $sanitize;
+    try {
+      $sanitize = $injector.get('$sanitize'); //tries to inject the sanitize service
+    }
+    catch(e) {
+      $log.warn('Unsafe use of typeahead please use ngSanitize'); // Warn the user about the danger
```
Comment (chrisirhc) [inline src/typeahead/typeahead.js]: Shouldn't log here, instead, log below only when the matchItem contains HTML. Users should only see this message if the labels contain HTML tags.

**Comment by chrisirhc** [comment]
Thanks @fernando-sendMail for your work on this, I did some further review.

**Comment by chrisirhc** [comment]
Also note, should remove dependency on `ui.bootstrap.bindHtml` module as it's no longer used after this change ( :+1: hooray!)

**Comment by fernando-sendMail** [comment]
@chrisirhc Thanks for the review, I'll be working on this over the week.

**Comment by fernando-sendMail** [comment]
Hey @chrisirhc I committed the latest stuff with the fixes and modifications to the test.

File: src/typeahead/typeahead.js [hunk 10]
```diff
@@ -399,13 +399,34 @@ angular.module('ui.bootstrap.typeahead', ['ui.bootstrap.position', 'ui.bootstrap
     };
   }])
 
-  .filter('typeaheadHighlight', function() {
+  .filter('typeaheadHighlight', ['$sce', '$injector', '$log', function($sce, $injector, $log) {
+
+    var $sanitize;
+    try {
+      $sanitize = $injector.get('$sanitize'); //tries to inject the sanitize service
```
Comment (chrisirhc) [inline src/typeahead/typeahead.js]: Since you're just checking for the existence, you can do `$injector.has('$sanitize')`. Also, probably can call this `hasSanitize` instead.

**Comment by chrisirhc** [comment]
Took another look. I saw that you updated `package.json` but the `angular-sanitize.js` is still in your PR and `karma.conf.js` hasn't been updated to use the angular-sanitize.js from node_modules.

**Comment by fernando-sendMail** [comment]
Hey @chrisirhc I think it's more convenient to name it ` isSanitizePresent ` given the use it has in this context.

I made a commit with the other changes that you noticed.

**Comment by chrisirhc** [comment]
Looks good to me. I'm scheduling this for 0.13.1 because we already have a lot of changes in 0.13.0 .

File: src/typeahead/test/typeahead-highlight-ngsanitize.spec.js [hunk 11]
```diff
@@ -0,0 +1,19 @@
+describe('Security concerns', function() {
+  var highlightFilter, 
+  $sanitize,
+  logSpy;
```
Comment (chrisirhc) [inline src/typeahead/test/typeahead-highlight-ngsanitize.spec.js]: Indentation

**Comment by fernando-sendMail** [comment]
Hey @chrisirhc It's been a while since I committed this.

Do I need to do something else?

**Comment by wesleycho** [comment]
This just needs to be rebased off of current `master` I believe - this LGTM, lot of good work here.

**Comment by fernando-sendMail** [comment]
Hey @wesleycho I'm having some trouble rebasing, I'm running:

`git rebase master`

and I resolve all the conflicts but, seems like it's the same result every time, is there a step that I'm missing?

**Comment by wesleycho** [comment]
Your merge commit might have messed things up a bit here - I will try to fix this locally and get this in.

**Comment by wesleycho** [comment]
Can you fix the broken test first?

**Comment by wesleycho** [comment]
@fernando-sendMail can you open a new PR with fixed history? It seems like something is broken with the history in this one, and I am having a bit of trouble untangling the conflicts.

**Comment by fernando-sendMail** [comment]
@wesleycho Sure, I'll work on that.

**Comment by fernando-sendMail** [comment]
Hey @wesleycho I think I resolved all the conflicts, the PR is ready for review again

**Comment by wesleycho** [comment]
Can you squash into one commit @fernando-sendMail ?

File: package.json [hunk 12]
```diff
@@ -1,38 +1,37 @@
 {
   "author": "https://github.com/angular-ui/bootstrap/graphs/contributors",
   "name": "angular-ui-bootstrap",
-  "version": "0.13.4-SNAPSHOT",
+  "version": "0.13.0-SNAPSHOT",
```
Comment (wesleycho) [inline package.json]: Change this back to 4

Comment (wesleycho) [inline package.json]: Revert all changes to this file excepting the addition of `angular-sanitize`

File: package.json [hunk 13]
```diff
@@ -1,38 +1,37 @@
...
   "repository": {
     "type": "git",
     "url": "https://github.com/angular-ui/bootstrap.git"
   },
   "devDependencies": {
-    "angular": "^1.4.3",
-    "angular-mocks": "^1.4.3",
+    "angular": "^1.3.15",
+    "angular-mocks": "^1.3.15",
+    "angular-sanitize": "^1.3.15",
```
Comment (wesleycho) [inline package.json]: Bump up to `^1.4.3`

File: template/typeahead/typeahead-match.html [hunk 14]
```diff
@@ -1 +1 @@
-<a href tabindex="-1" bind-html-unsafe="match.label | typeaheadHighlight:query"></a>
+<a tabindex="-1" ng-bind-html="match.label | typeaheadHighlight:query"></a>
```
Comment (wesleycho) [inline template/typeahead/typeahead-match.html]: Keep the empty `href` - this is for accessibility.

**Comment by wesleycho** [comment]
With these minor reverts/changes, this should be good to merge. Squashing the commits makes it easier to replay the change on top of `master` when we merge it, which is very useful for us.

**Comment by fernando-sendMail** [comment]
Hey @wesleycho I tried using `git rebase -i <commit to squash>` but that didn't worked for me, it created an aditional commit instead, am I doing something wrong?

**Comment by wesleycho** [comment]
You can do `git rebase -i HEAD~[number of commits without brackets]`

**Comment by wesleycho** [comment]
Keep in mind once you do this, you will have to force push to the remote to overwrite the history with the new version.

**Comment by fernando-sendMail** [comment]
@wesleycho Finally I've made it, it's ready for review again.

**Comment by wesleycho** [comment]
Thanks a lot for sticking with it through the time, this is good work!

**Comment by fernando-sendMail** [comment]
No problem, thanks a lot for the support guys!
